<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Jonathan Dreyfuss" />

<meta name="date" content="2019-01-28" />

<title>ezlimma vignette</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' || rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">ezlimma vignette</h1>
<h4 class="author"><em>Jonathan Dreyfuss</em></h4>
<h4 class="date"><em>2019-01-28</em></h4>



<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>Welcome to <code>ezlimma</code>. This package is a wrapper for the commonly used <code>limma</code> package, which makes <code>limma</code> easier to use and offers some functionality that adds value to <code>limma</code>’s functions, such as mediation analysis and allowing for writing pathway results to Excel workbooks that link to the feature-level (e.g. gene-level) results. It exposes some of <code>limma</code>’s parameters, such as those in <code>lmFit</code>; when the parameters are not exposed, <code>limma</code>’s default is used.</p>
</div>
<div id="install-load-package" class="section level2">
<h2>Install &amp; load package</h2>
<p>See README.md.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(limma)
<span class="kw">library</span>(ezlimma)</code></pre></div>
</div>
<div id="simulate-data" class="section level2">
<h2>Simulate data</h2>
<p>To demonstrate the package on a dataset, I’ll simulate log2 expression data with 100 genes and 6 samples in two groups with a continuous covariate. The covariate and the first gene will be associated with group. This is similar to the example data from <code>limma</code> <code>contrasts.fit</code>.</p>
<p>We treat this data as being already processed. For many datasets, there are zeroes or missing values that might need to be imputed; samples need to be normalized to make them comparable and amenable to statistical tests; absent genes need to be removed; sample outliers need to be assessed to examine whether some experimental variables or batch effects need to be accounted for, or the samples need to be removed or down-weighted; and trends between a gene’s mean expression and its variance should be accounted for, especially in RNA-seq data, with <code>limma</code>’s <code>voom</code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">42</span>)
M &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rnorm</span>(<span class="dv">100</span><span class="op">*</span><span class="dv">6</span>, <span class="dt">sd=</span><span class="fl">0.3</span>), <span class="dt">nrow=</span><span class="dv">100</span>, <span class="dt">ncol=</span><span class="dv">6</span>)
<span class="kw">dimnames</span>(M) &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="kw">paste0</span>(<span class="st">&quot;gene&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(M)), <span class="kw">paste0</span>(<span class="st">&quot;sample&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(M)))
M[<span class="dv">1</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>] &lt;-<span class="st"> </span>M[<span class="dv">1</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>] <span class="op">+</span><span class="st"> </span><span class="dv">2</span>

pheno &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">grp=</span><span class="kw">rep</span>(<span class="kw">c</span>(<span class="st">&quot;First3&quot;</span>, <span class="st">&quot;Last3&quot;</span>), <span class="dt">each=</span><span class="dv">3</span>), <span class="dt">yy=</span><span class="kw">rnorm</span>(<span class="kw">ncol</span>(M)))
pheno<span class="op">$</span>yy[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>] &lt;-<span class="st"> </span>pheno<span class="op">$</span>yy[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]<span class="op">-</span><span class="dv">2</span></code></pre></div>
</div>
<div id="gene-level-analysis" class="section level2">
<h2>Gene level analysis</h2>
<p>Assume we want to test if genes are changed in <code>First3</code>, in <code>Last3</code>, or in <code>Last3 - First3</code>. We can do this in two lines of code,</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">contr.v &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dt">First3=</span><span class="st">&quot;First3&quot;</span>, <span class="dt">Last3=</span><span class="st">&quot;Last3&quot;</span>, <span class="dt">Last3vsFirst3=</span><span class="st">&quot;Last3-First3&quot;</span>)
res0 &lt;-<span class="st"> </span><span class="kw">limma_contrasts</span>(M, <span class="dt">grp=</span>pheno<span class="op">$</span>grp, <span class="dt">contrast.v =</span> contr.v)</code></pre></div>
<p><code>First3</code> tests if the mean expression of genes in these samples differs from zero, and <code>Last3vsFirst3</code> tests difference between groups.</p>
<p>This returns a data frame that looks like:</p>
<table>
<thead>
<tr class="header">
<th></th>
<th align="right">First3.avg</th>
<th align="right">Last3.avg</th>
<th align="right">First3.p</th>
<th align="right">First3.FDR</th>
<th align="right">First3.logFC</th>
<th align="right">First3.FC</th>
<th align="right">Last3.p</th>
<th align="right">Last3.FDR</th>
<th align="right">Last3.logFC</th>
<th align="right">Last3.FC</th>
<th align="right">Last3vsFirst3.p</th>
<th align="right">Last3vsFirst3.FDR</th>
<th align="right">Last3vsFirst3.logFC</th>
<th align="right">Last3vsFirst3.FC</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>gene1</td>
<td align="right">2.0600</td>
<td align="right">0.236</td>
<td align="right">0.00000</td>
<td align="right">0.000</td>
<td align="right">2.0600</td>
<td align="right">4.16</td>
<td align="right">0.156000</td>
<td align="right">0.741</td>
<td align="right">0.236</td>
<td align="right">1.18</td>
<td align="right">0.000000</td>
<td align="right">0.0000</td>
<td align="right">-1.820</td>
<td align="right">-3.53</td>
</tr>
<tr class="even">
<td>gene25</td>
<td align="right">0.2320</td>
<td align="right">0.593</td>
<td align="right">0.16300</td>
<td align="right">0.987</td>
<td align="right">0.2320</td>
<td align="right">1.17</td>
<td align="right">0.000401</td>
<td align="right">0.020</td>
<td align="right">0.593</td>
<td align="right">1.51</td>
<td align="right">0.125000</td>
<td align="right">0.7080</td>
<td align="right">0.360</td>
<td align="right">1.28</td>
</tr>
<tr class="odd">
<td>gene74</td>
<td align="right">0.1170</td>
<td align="right">-0.609</td>
<td align="right">0.48200</td>
<td align="right">0.987</td>
<td align="right">0.1170</td>
<td align="right">1.08</td>
<td align="right">0.000278</td>
<td align="right">0.020</td>
<td align="right">-0.609</td>
<td align="right">-1.53</td>
<td align="right">0.002140</td>
<td align="right">0.0713</td>
<td align="right">-0.726</td>
<td align="right">-1.65</td>
</tr>
<tr class="even">
<td>gene59</td>
<td align="right">-0.4540</td>
<td align="right">0.359</td>
<td align="right">0.00648</td>
<td align="right">0.198</td>
<td align="right">-0.4540</td>
<td align="right">-1.37</td>
<td align="right">0.031400</td>
<td align="right">0.431</td>
<td align="right">0.359</td>
<td align="right">1.28</td>
<td align="right">0.000593</td>
<td align="right">0.0297</td>
<td align="right">0.813</td>
<td align="right">1.76</td>
</tr>
<tr class="odd">
<td>gene17</td>
<td align="right">-0.0929</td>
<td align="right">-0.469</td>
<td align="right">0.57600</td>
<td align="right">0.987</td>
<td align="right">-0.0929</td>
<td align="right">-1.07</td>
<td align="right">0.004930</td>
<td align="right">0.164</td>
<td align="right">-0.469</td>
<td align="right">-1.38</td>
<td align="right">0.110000</td>
<td align="right">0.7080</td>
<td align="right">-0.376</td>
<td align="right">-1.30</td>
</tr>
<tr class="even">
<td>gene12</td>
<td align="right">0.4430</td>
<td align="right">-0.148</td>
<td align="right">0.00791</td>
<td align="right">0.198</td>
<td align="right">0.4430</td>
<td align="right">1.36</td>
<td align="right">0.372000</td>
<td align="right">0.917</td>
<td align="right">-0.148</td>
<td align="right">-1.11</td>
<td align="right">0.012200</td>
<td align="right">0.2430</td>
<td align="right">-0.591</td>
<td align="right">-1.51</td>
</tr>
<tr class="odd">
<td>gene4</td>
<td align="right">0.4540</td>
<td align="right">0.092</td>
<td align="right">0.00651</td>
<td align="right">0.198</td>
<td align="right">0.4540</td>
<td align="right">1.37</td>
<td align="right">0.580000</td>
<td align="right">0.917</td>
<td align="right">0.092</td>
<td align="right">1.07</td>
<td align="right">0.124000</td>
<td align="right">0.7080</td>
<td align="right">-0.362</td>
<td align="right">-1.29</td>
</tr>
<tr class="even">
<td>gene66</td>
<td align="right">0.3800</td>
<td align="right">-0.229</td>
<td align="right">0.02260</td>
<td align="right">0.323</td>
<td align="right">0.3800</td>
<td align="right">1.30</td>
<td align="right">0.169000</td>
<td align="right">0.741</td>
<td align="right">-0.229</td>
<td align="right">-1.17</td>
<td align="right">0.009890</td>
<td align="right">0.2430</td>
<td align="right">-0.609</td>
<td align="right">-1.52</td>
</tr>
<tr class="odd">
<td>gene13</td>
<td align="right">-0.1770</td>
<td align="right">-0.383</td>
<td align="right">0.28800</td>
<td align="right">0.987</td>
<td align="right">-0.1770</td>
<td align="right">-1.13</td>
<td align="right">0.021700</td>
<td align="right">0.412</td>
<td align="right">-0.383</td>
<td align="right">-1.30</td>
<td align="right">0.381000</td>
<td align="right">0.8940</td>
<td align="right">-0.206</td>
<td align="right">-1.15</td>
</tr>
<tr class="even">
<td>gene60</td>
<td align="right">0.1190</td>
<td align="right">-0.395</td>
<td align="right">0.47400</td>
<td align="right">0.987</td>
<td align="right">0.1190</td>
<td align="right">1.09</td>
<td align="right">0.017800</td>
<td align="right">0.412</td>
<td align="right">-0.395</td>
<td align="right">-1.32</td>
<td align="right">0.029200</td>
<td align="right">0.4170</td>
<td align="right">-0.514</td>
<td align="right">-1.43</td>
</tr>
</tbody>
</table>
<p>We can test association to the continuous covariate with:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">res1 &lt;-<span class="st"> </span><span class="kw">ezcor</span>(M, pheno<span class="op">$</span>yy, <span class="dt">method=</span><span class="st">&quot;pearson&quot;</span>)</code></pre></div>
<p>where <code>method</code> can be any of <code>&quot;pearson&quot;</code>, <code>&quot;kendall&quot;</code>, or <code>&quot;spearman&quot;</code>. This returns a data frame like,</p>
<table>
<thead>
<tr class="header">
<th></th>
<th align="right">cor</th>
<th align="right">p</th>
<th align="right">FDR</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>gene66</td>
<td align="right">-0.925</td>
<td align="right">0.00826</td>
<td align="right">0.754</td>
</tr>
<tr class="even">
<td>gene1</td>
<td align="right">-0.875</td>
<td align="right">0.02250</td>
<td align="right">0.754</td>
</tr>
<tr class="odd">
<td>gene60</td>
<td align="right">-0.860</td>
<td align="right">0.02790</td>
<td align="right">0.754</td>
</tr>
<tr class="even">
<td>gene22</td>
<td align="right">0.789</td>
<td align="right">0.06210</td>
<td align="right">0.754</td>
</tr>
<tr class="odd">
<td>gene9</td>
<td align="right">-0.772</td>
<td align="right">0.07170</td>
<td align="right">0.754</td>
</tr>
<tr class="even">
<td>gene80</td>
<td align="right">0.768</td>
<td align="right">0.07470</td>
<td align="right">0.754</td>
</tr>
<tr class="odd">
<td>gene47</td>
<td align="right">0.765</td>
<td align="right">0.07620</td>
<td align="right">0.754</td>
</tr>
<tr class="even">
<td>gene19</td>
<td align="right">0.759</td>
<td align="right">0.08000</td>
<td align="right">0.754</td>
</tr>
<tr class="odd">
<td>gene99</td>
<td align="right">-0.757</td>
<td align="right">0.08140</td>
<td align="right">0.754</td>
</tr>
<tr class="even">
<td>gene15</td>
<td align="right">0.755</td>
<td align="right">0.08260</td>
<td align="right">0.754</td>
</tr>
</tbody>
</table>
<p>Similarly, we can test association using linear regression with <code>limma</code>, where the model is <em>gene expression = a + bx + error</em>, and we test coefficient 2, i.e. the null hypothesis that <em>b=0</em>. So it returns slopes rather than correlation coefficients. The code is,</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">res2 &lt;-<span class="st"> </span><span class="kw">limma_cor</span>(M, <span class="dt">phenotype =</span> pheno<span class="op">$</span>yy)</code></pre></div>
<p>The output is,</p>
<table>
<thead>
<tr class="header">
<th></th>
<th align="right">AveExpr</th>
<th align="right">p</th>
<th align="right">FDR</th>
<th align="right">slope</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>gene1</td>
<td align="right">1.1500</td>
<td align="right">0.00000</td>
<td align="right">0.000</td>
<td align="right">-0.644</td>
</tr>
<tr class="even">
<td>gene66</td>
<td align="right">0.0757</td>
<td align="right">0.00536</td>
<td align="right">0.268</td>
<td align="right">-0.250</td>
</tr>
<tr class="odd">
<td>gene9</td>
<td align="right">0.0285</td>
<td align="right">0.01340</td>
<td align="right">0.314</td>
<td align="right">-0.222</td>
</tr>
<tr class="even">
<td>gene19</td>
<td align="right">-0.2180</td>
<td align="right">0.01930</td>
<td align="right">0.314</td>
<td align="right">0.210</td>
</tr>
<tr class="odd">
<td>gene59</td>
<td align="right">-0.0479</td>
<td align="right">0.02030</td>
<td align="right">0.314</td>
<td align="right">0.208</td>
</tr>
<tr class="even">
<td>gene15</td>
<td align="right">0.0376</td>
<td align="right">0.02050</td>
<td align="right">0.314</td>
<td align="right">0.208</td>
</tr>
<tr class="odd">
<td>gene99</td>
<td align="right">-0.0156</td>
<td align="right">0.02680</td>
<td align="right">0.314</td>
<td align="right">-0.199</td>
</tr>
<tr class="even">
<td>gene60</td>
<td align="right">-0.1380</td>
<td align="right">0.02720</td>
<td align="right">0.314</td>
<td align="right">-0.198</td>
</tr>
<tr class="odd">
<td>gene12</td>
<td align="right">0.1470</td>
<td align="right">0.02820</td>
<td align="right">0.314</td>
<td align="right">-0.197</td>
</tr>
<tr class="even">
<td>gene39</td>
<td align="right">-0.1610</td>
<td align="right">0.03870</td>
<td align="right">0.387</td>
<td align="right">0.186</td>
</tr>
</tbody>
</table>
<p>Because this is a regression, we can test for association to <code>yy</code> while adjusting for <code>grp</code>. We do this by creating our own design matrix instead of letting <code>limma_cor</code> create one from <code>phenotype</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">design &lt;-<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span><span class="dv">1</span><span class="op">+</span>yy<span class="op">+</span>grp, <span class="dt">data=</span>pheno)
res3 &lt;-<span class="st"> </span><span class="kw">limma_cor</span>(M, <span class="dt">design =</span> design)</code></pre></div>
<p>which gives</p>
<table>
<thead>
<tr class="header">
<th></th>
<th align="right">AveExpr</th>
<th align="right">p</th>
<th align="right">FDR</th>
<th align="right">slope</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>gene9</td>
<td align="right">0.02850</td>
<td align="right">0.00383</td>
<td align="right">0.305</td>
<td align="right">-0.451</td>
</tr>
<tr class="even">
<td>gene31</td>
<td align="right">0.08570</td>
<td align="right">0.00904</td>
<td align="right">0.305</td>
<td align="right">-0.406</td>
</tr>
<tr class="odd">
<td>gene98</td>
<td align="right">0.02890</td>
<td align="right">0.00914</td>
<td align="right">0.305</td>
<td align="right">0.406</td>
</tr>
<tr class="even">
<td>gene39</td>
<td align="right">-0.16100</td>
<td align="right">0.01360</td>
<td align="right">0.339</td>
<td align="right">0.384</td>
</tr>
<tr class="odd">
<td>gene15</td>
<td align="right">0.03760</td>
<td align="right">0.03500</td>
<td align="right">0.537</td>
<td align="right">0.328</td>
</tr>
<tr class="even">
<td>gene40</td>
<td align="right">-0.05700</td>
<td align="right">0.03770</td>
<td align="right">0.537</td>
<td align="right">-0.323</td>
</tr>
<tr class="odd">
<td>gene44</td>
<td align="right">0.00594</td>
<td align="right">0.03820</td>
<td align="right">0.537</td>
<td align="right">0.322</td>
</tr>
<tr class="even">
<td>gene37</td>
<td align="right">-0.15400</td>
<td align="right">0.04560</td>
<td align="right">0.537</td>
<td align="right">0.310</td>
</tr>
<tr class="odd">
<td>gene19</td>
<td align="right">-0.21800</td>
<td align="right">0.05260</td>
<td align="right">0.537</td>
<td align="right">0.301</td>
</tr>
<tr class="even">
<td>gene34</td>
<td align="right">-0.00532</td>
<td align="right">0.05450</td>
<td align="right">0.537</td>
<td align="right">0.299</td>
</tr>
</tbody>
</table>
<p>We can also test multiple phenotypes together. Say we had a matrix of phenotypes as</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pheno.mat &lt;-<span class="st"> </span><span class="kw">cbind</span>(<span class="dt">pheno1=</span>pheno<span class="op">$</span>yy, <span class="dt">pheno2=</span><span class="kw">rnorm</span>(<span class="kw">ncol</span>(M)))</code></pre></div>
<p>then we could test both phenotypes together in one line, as</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">res4 &lt;-<span class="st"> </span><span class="kw">multi_cor</span>(<span class="dt">object=</span>M, <span class="dt">pheno.tab =</span> pheno.mat, <span class="dt">reorder.rows =</span> <span class="ot">TRUE</span>)</code></pre></div>
<table>
<thead>
<tr class="header">
<th></th>
<th align="right">pheno1.cor</th>
<th align="right">pheno1.p</th>
<th align="right">pheno1.FDR</th>
<th align="right">pheno2.cor</th>
<th align="right">pheno2.p</th>
<th align="right">pheno2.FDR</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>gene60</td>
<td align="right">-0.860</td>
<td align="right">0.0279</td>
<td align="right">0.754</td>
<td align="right">0.417</td>
<td align="right">0.4110</td>
<td align="right">0.964</td>
</tr>
<tr class="even">
<td>gene39</td>
<td align="right">0.701</td>
<td align="right">0.1210</td>
<td align="right">0.776</td>
<td align="right">0.634</td>
<td align="right">0.1760</td>
<td align="right">0.933</td>
</tr>
<tr class="odd">
<td>gene67</td>
<td align="right">0.500</td>
<td align="right">0.3120</td>
<td align="right">0.855</td>
<td align="right">-0.798</td>
<td align="right">0.0573</td>
<td align="right">0.933</td>
</tr>
<tr class="even">
<td>gene98</td>
<td align="right">0.602</td>
<td align="right">0.2060</td>
<td align="right">0.855</td>
<td align="right">0.719</td>
<td align="right">0.1080</td>
<td align="right">0.933</td>
</tr>
<tr class="odd">
<td>gene1</td>
<td align="right">-0.875</td>
<td align="right">0.0225</td>
<td align="right">0.754</td>
<td align="right">0.356</td>
<td align="right">0.4880</td>
<td align="right">0.982</td>
</tr>
<tr class="even">
<td>gene86</td>
<td align="right">-0.502</td>
<td align="right">0.3100</td>
<td align="right">0.855</td>
<td align="right">-0.723</td>
<td align="right">0.1040</td>
<td align="right">0.933</td>
</tr>
<tr class="odd">
<td>gene34</td>
<td align="right">0.397</td>
<td align="right">0.4360</td>
<td align="right">0.890</td>
<td align="right">0.786</td>
<td align="right">0.0639</td>
<td align="right">0.933</td>
</tr>
<tr class="even">
<td>gene19</td>
<td align="right">0.759</td>
<td align="right">0.0800</td>
<td align="right">0.754</td>
<td align="right">0.420</td>
<td align="right">0.4070</td>
<td align="right">0.964</td>
</tr>
<tr class="odd">
<td>gene74</td>
<td align="right">-0.551</td>
<td align="right">0.2580</td>
<td align="right">0.855</td>
<td align="right">0.649</td>
<td align="right">0.1630</td>
<td align="right">0.933</td>
</tr>
<tr class="even">
<td>gene38</td>
<td align="right">-0.471</td>
<td align="right">0.3460</td>
<td align="right">0.855</td>
<td align="right">0.712</td>
<td align="right">0.1120</td>
<td align="right">0.933</td>
</tr>
</tbody>
</table>
<p>where the result is ordered by the combined p-value of the two associations per gene. The <code>method</code> options for the association test are: <code>&quot;pearson&quot;</code> (the default), <code>&quot;spearman&quot;</code>, <code>&quot;kendall&quot;</code>, as before, or <code>&quot;limma&quot;</code>, where it will use <code>limma_cor</code>. This function only allows one method for all phenotypes.</p>
<p>We can also test what genes mediate the effect of group on the continuous phenotype. This is called high-throughput mediation analysis (<code>hitman</code>). We set the exposure (<code>E</code>) to group, the outcome (<code>Y</code>) to <code>pheno.v</code>, and test each gene in <code>M</code> as a mediator. For <code>hitman</code> we transform the exposure to numeric, and set the names of the <code>phenoype</code> to those of <code>M</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">grp.num &lt;-<span class="st"> </span><span class="kw">batch2design</span>(pheno<span class="op">$</span>grp)
res.hit &lt;-<span class="st"> </span><span class="kw">hitman</span>(<span class="dt">E =</span> grp.num, <span class="dt">M =</span> M, <span class="dt">Y =</span> <span class="kw">setNames</span>(pheno<span class="op">$</span>yy, <span class="dt">nm=</span><span class="kw">colnames</span>(M)))</code></pre></div>
<p><code>hitman</code> returns a table with a combined (“EMY”) p-value and FDR for each gene. This combined statistic is based on the association of E to the gene in <code>M</code>, and the conditional association of the gene in <code>M</code> to <code>Y</code> given <code>E</code>.</p>
<table>
<thead>
<tr class="header">
<th></th>
<th align="right">EMY.p</th>
<th align="right">EMY.FDR</th>
<th align="right">EM_dir.p</th>
<th align="right">MY_dir.p</th>
<th align="right">EM.p</th>
<th align="right">EM.t</th>
<th align="right">MY.p</th>
<th align="right">MY.slope</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>gene1</td>
<td align="right">0.00545</td>
<td align="right">0.477</td>
<td align="right">0.00000</td>
<td align="right">0.073800</td>
<td align="right">0.00000</td>
<td align="right">7.76</td>
<td align="right">0.073800</td>
<td align="right">-0.240</td>
</tr>
<tr class="even">
<td>gene19</td>
<td align="right">0.01380</td>
<td align="right">0.477</td>
<td align="right">0.11700</td>
<td align="right">0.025400</td>
<td align="right">0.11700</td>
<td align="right">-1.57</td>
<td align="right">0.025400</td>
<td align="right">0.301</td>
</tr>
<tr class="odd">
<td>gene99</td>
<td align="right">0.01610</td>
<td align="right">0.477</td>
<td align="right">0.06800</td>
<td align="right">0.127000</td>
<td align="right">0.06800</td>
<td align="right">1.83</td>
<td align="right">0.127000</td>
<td align="right">-0.205</td>
</tr>
<tr class="even">
<td>gene66</td>
<td align="right">0.02360</td>
<td align="right">0.477</td>
<td align="right">0.00989</td>
<td align="right">0.154000</td>
<td align="right">0.00989</td>
<td align="right">2.59</td>
<td align="right">0.154000</td>
<td align="right">-0.192</td>
</tr>
<tr class="odd">
<td>gene15</td>
<td align="right">0.02380</td>
<td align="right">0.477</td>
<td align="right">0.15400</td>
<td align="right">0.015000</td>
<td align="right">0.15400</td>
<td align="right">-1.43</td>
<td align="right">0.015000</td>
<td align="right">0.328</td>
</tr>
<tr class="even">
<td>gene80</td>
<td align="right">0.03810</td>
<td align="right">0.635</td>
<td align="right">0.19500</td>
<td align="right">0.118000</td>
<td align="right">0.19500</td>
<td align="right">-1.30</td>
<td align="right">0.118000</td>
<td align="right">0.210</td>
</tr>
<tr class="odd">
<td>gene48</td>
<td align="right">0.07680</td>
<td align="right">0.820</td>
<td align="right">0.12700</td>
<td align="right">0.277000</td>
<td align="right">0.12700</td>
<td align="right">1.53</td>
<td align="right">0.277000</td>
<td align="right">-0.146</td>
</tr>
<tr class="even">
<td>gene9</td>
<td align="right">0.07720</td>
<td align="right">0.820</td>
<td align="right">0.27800</td>
<td align="right">0.000864</td>
<td align="right">0.27800</td>
<td align="right">1.09</td>
<td align="right">0.000864</td>
<td align="right">-0.451</td>
</tr>
<tr class="odd">
<td>gene47</td>
<td align="right">0.08090</td>
<td align="right">0.820</td>
<td align="right">0.28400</td>
<td align="right">0.239000</td>
<td align="right">0.28400</td>
<td align="right">-1.07</td>
<td align="right">0.239000</td>
<td align="right">0.158</td>
</tr>
<tr class="even">
<td>gene25</td>
<td align="right">0.08200</td>
<td align="right">0.820</td>
<td align="right">0.12500</td>
<td align="right">0.286000</td>
<td align="right">0.12500</td>
<td align="right">-1.54</td>
<td align="right">0.286000</td>
<td align="right">0.143</td>
</tr>
</tbody>
</table>
</div>
<div id="gene-set-level-analysis" class="section level2">
<h2>Gene set level analysis</h2>
<p>At a higher level than features (e.g. genes) lie feature sets, such as gene sets or pathways. These can be tested with <code>limma</code>’s function <code>roast</code>, which are implemented for multiple sets at a time with the functions <code>mroast</code> or its faster approximation <code>fry</code>. We have written wrappers around these to allow for easier testing, and to make an Excel file where each gene set links to a CSV containing the statistics on its genes, made using the feature-level functions.</p>
<p>Let’s make the gene set object,</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">G &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="kw">list</span>(<span class="dt">name=</span><span class="st">&quot;pwy1&quot;</span>, <span class="dt">description=</span><span class="ot">NA</span>, <span class="dt">genes=</span><span class="kw">paste0</span>(<span class="st">&quot;gene&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>)),
          <span class="kw">list</span>(<span class="dt">name=</span><span class="st">&quot;pwy2&quot;</span>, <span class="dt">description=</span><span class="ot">NA</span>, <span class="dt">genes=</span><span class="kw">paste0</span>(<span class="st">&quot;gene&quot;</span>, <span class="dv">11</span><span class="op">:</span><span class="dv">20</span>)),
          <span class="kw">list</span>(<span class="dt">name=</span><span class="st">&quot;pwy3&quot;</span>, <span class="dt">description=</span><span class="ot">NA</span>, <span class="dt">genes=</span><span class="kw">paste0</span>(<span class="st">&quot;gene&quot;</span>, <span class="dv">21</span><span class="op">:</span><span class="dv">30</span>)))</code></pre></div>
<p>This object would normally be read in using <code>read_gmt</code>. Then we can test each of these gene sets.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">res5 &lt;-<span class="st"> </span><span class="kw">roast_contrasts</span>(<span class="dt">object=</span>M, <span class="dt">G=</span>G, <span class="dt">feat.tab=</span>res0, <span class="dt">grp=</span>pheno<span class="op">$</span>grp, <span class="dt">contrast.v =</span> contr.v, <span class="dt">fun=</span><span class="st">&quot;fry&quot;</span>, <span class="dt">name=</span><span class="ot">NA</span>)</code></pre></div>
<p>If we set <code>name</code> to a value other than <code>NA</code>, this would create a folder with one CSV per pathway, and an Excel file that looks like</p>
<table>
<thead>
<tr class="header">
<th></th>
<th align="right">NGenes</th>
<th align="left">First3.Direction</th>
<th align="right">First3.PropUpP05</th>
<th align="right">First3.PropDownP05</th>
<th align="right">First3.p</th>
<th align="right">First3.FDR</th>
<th align="right">First3.Mixed.p</th>
<th align="right">First3.Mixed.FDR</th>
<th align="left">Last3.Direction</th>
<th align="right">Last3.PropUpP05</th>
<th align="right">Last3.PropDownP05</th>
<th align="right">Last3.p</th>
<th align="right">Last3.FDR</th>
<th align="right">Last3.Mixed.p</th>
<th align="right">Last3.Mixed.FDR</th>
<th align="left">Last3vsFirst3.Direction</th>
<th align="right">Last3vsFirst3.PropUpP05</th>
<th align="right">Last3vsFirst3.PropDownP05</th>
<th align="right">Last3vsFirst3.p</th>
<th align="right">Last3vsFirst3.FDR</th>
<th align="right">Last3vsFirst3.Mixed.p</th>
<th align="right">Last3vsFirst3.Mixed.FDR</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>pwy1</td>
<td align="right">10</td>
<td align="left">Up</td>
<td align="right">0.2</td>
<td align="right">0.0</td>
<td align="right">0.0147405</td>
<td align="right">0.0442216</td>
<td align="right">0.0003846</td>
<td align="right">0.0011539</td>
<td align="left">Down</td>
<td align="right">0.0</td>
<td align="right">0.0</td>
<td align="right">0.9380988</td>
<td align="right">0.9380988</td>
<td align="right">0.8452777</td>
<td align="right">0.8452777</td>
<td align="left">Down</td>
<td align="right">0</td>
<td align="right">0.1</td>
<td align="right">0.0413832</td>
<td align="right">0.1241496</td>
<td align="right">0.0042642</td>
<td align="right">0.0127927</td>
</tr>
<tr class="even">
<td>pwy3</td>
<td align="right">10</td>
<td align="left">Down</td>
<td align="right">0.0</td>
<td align="right">0.2</td>
<td align="right">0.1277691</td>
<td align="right">0.1916536</td>
<td align="right">0.0771335</td>
<td align="right">0.1157002</td>
<td align="left">Up</td>
<td align="right">0.1</td>
<td align="right">0.0</td>
<td align="right">0.5373521</td>
<td align="right">0.8060281</td>
<td align="right">0.1031779</td>
<td align="right">0.3095336</td>
<td align="left">Up</td>
<td align="right">0</td>
<td align="right">0.0</td>
<td align="right">0.1409461</td>
<td align="right">0.2114191</td>
<td align="right">0.3117519</td>
<td align="right">0.3849170</td>
</tr>
<tr class="odd">
<td>pwy2</td>
<td align="right">10</td>
<td align="left">Down</td>
<td align="right">0.1</td>
<td align="right">0.1</td>
<td align="right">0.7183857</td>
<td align="right">0.7183857</td>
<td align="right">0.3703035</td>
<td align="right">0.3703035</td>
<td align="left">Down</td>
<td align="right">0.0</td>
<td align="right">0.2</td>
<td align="right">0.3875927</td>
<td align="right">0.8060281</td>
<td align="right">0.2362813</td>
<td align="right">0.3544220</td>
<td align="left">Down</td>
<td align="right">0</td>
<td align="right">0.1</td>
<td align="right">0.7020194</td>
<td align="right">0.7020194</td>
<td align="right">0.3849170</td>
<td align="right">0.3849170</td>
</tr>
</tbody>
</table>
<p>except the first column would have links to the CSVs in the folder. Similarly, we can use <code>roast_cor</code>.</p>
</div>
<div id="system-info" class="section level2">
<h2>System info</h2>
<p>Here is the information on the system on which this document was compiled:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>## R version 3.5.2 (2018-12-20)
## Platform: x86_64-w64-mingw32/x64 (64-bit)
## Running under: Windows 7 x64 (build 7601) Service Pack 1
## 
## Matrix products: default
## 
## locale:
## [1] LC_COLLATE=English_United States.1252 
## [2] LC_CTYPE=English_United States.1252   
## [3] LC_MONETARY=English_United States.1252
## [4] LC_NUMERIC=C                          
## [5] LC_TIME=English_United States.1252    
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] ezlimma_0.2.1.9002 limma_3.36.5      
## 
## loaded via a namespace (and not attached):
##  [1] compiler_3.5.2  magrittr_1.5    tools_3.5.2     htmltools_0.3.6
##  [5] yaml_2.2.0      Rcpp_1.0.0      stringi_1.2.4   rmarkdown_1.11 
##  [9] highr_0.7       knitr_1.21      stringr_1.3.1   xfun_0.4       
## [13] digest_0.6.18   statmod_1.4.30  evaluate_0.12</code></pre>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
